### PR TITLE
This PR introduces the Summarized Email Report with Summary Graphs

### LINKED ISSUE #910: Modify auto-generating email template
**Description:**

In order to increase the visual impact and the understandability of representing data, modify the auto-generating email and add the following graphs and tables into the mail.

- Test execution history graph: This graph represents the summary of previously executed build jobs.
- Last build summary graph: This graph represents the summary of the last build.
- Test failure summary table: This table represents the failed test cases against infra combinations.

In order to complete this task, it is divided into two subtasks. Those are 
- Implementing back-end logic to get data to draw these graphs and tables
- Implementing report generating and graph drawing method

### LINKED ISSUE #911: Implementing back-end logic to draw graphs and tables [auto-generating mail]
**Description:**

Please refer  #910

### LINKED ISSUE #912: Implementing report generating and graph drawing method [auto-generating mail]
**Description:**

Please refer #910

### PR DESCRIPTION
**Pull request by yasassri**
**Purpose**
This resolves #910 and resolves #911 and resolves #912 

Included graphs are as follows,

history

summary

**Goals**

To generate more condensed report that is user friendly.

**Approach**

Chart generation used JavaFx for rendering graphs.

### CODE REVIEW

File: common/src/main/java/org/wso2/testgrid/common/InfraCombination.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+package org.wso2.testgrid.common;
+
+/**
+ * This defines s single infra combination.
+ *
+ * @since 1.0.0
+ */
+public class InfraCombination {
+    private String os;
```
Comment (kasunbg) [inline common/src/main/java/org/wso2/testgrid/common/InfraCombination.java]: We already have an InfrastructureCombination in place. We need to reuse existing stuff!

File: common/src/main/java/org/wso2/testgrid/common/util/FileUtil.java [hunk 2]
```diff
@@ -162,4 +170,83 @@ public static void compress(String sourceDir, String destination) throws IOExcep
...
+
+    /**
+     * Get test plan ids by reading testgrid yaml files contains in the testgrid home.
+     *
+     * @param workspace path of the current workspace
+     * @throws TestGridException thrown when IO exception on reading testgrid yaml files.
+     */
+    public static List<String> getTestPlanIdByReadingTGYaml(String workspace) throws TestGridException {
+        List<String> testPlanIds = new ArrayList<>();
+        Path source = Paths.get(workspace, "test-plans");
```
Comment (kasunbg) [inline common/src/main/java/org/wso2/testgrid/common/util/FileUtil.java]: We have a constant for test-plans.

File: common/src/main/java/org/wso2/testgrid/common/InfraCombination.java (hunk 1 above)
Comment (lasanthaDLPDS) [reply common/src/main/java/org/wso2/testgrid/common/InfraCombination.java]: Yes, we have an InfrastructureCombination but in order to generate the graph, we used a simplified version. It is used to maintain OS, JDK and DB Engine.

File: dao/src/main/java/org/wso2/testgrid/dao/repository/TestPlanRepository.java [hunk 3]
```diff
@@ -290,4 +295,120 @@ public TestPlan getLastFailure(TestPlan testPlan) {
...
+        return getResultList(query, TestCaseFailureResultDTO.class);
+    }
+
+    /**
+     * This method returns the test execution summary for given test plan ids(i.e for a given build job).
+     *
+     * @param testPlanIds test plan ids of a specific build job
+     * @return a List of {@link String} representing statuses of given test plans
+     */
+    public List<String> getTestExecutionSummaryByTPId(List<String> testPlanIds) {
```
Comment (kasunbg) [inline dao/src/main/java/org/wso2/testgrid/dao/repository/TestPlanRepository.java]: As I understand, this is a list of Statuses. Shall we use a method name like getTestPlanStatuses instead?

Comment (kasunbg) [reply dao/src/main/java/org/wso2/testgrid/dao/repository/TestPlanRepository.java]: Also, it would make much sense if we return a List<Status>.

File: reporting/src/main/java/org/wso2/testgrid/reporting/GraphDataProvider.java [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+            Gson gson = new GsonBuilder().create();
+            InfraCombination infraCombination = new InfraCombination();
+            String testName = testFailure.getName();
+            JsonElement jelem = gson.fromJson(testFailure.getInfraParameters(), JsonElement.class);
+            JsonObject jobj = jelem.getAsJsonObject();
+            infraCombination.setOs(StringUtil
+                    .concatStrings(jobj.get("OS").getAsString(), " - ", jobj.get("OSVersion").getAsString()));
+            infraCombination.setJdk(jobj.get("JDK").getAsString());
+            infraCombination.setDbEngine(StringUtil.concatStrings(jobj.get("DBEngine").getAsString(), " - ",
+                    jobj.get("DBEngineVersion").getAsString()));
```
Comment (harshanL) [inline reporting/src/main/java/org/wso2/testgrid/reporting/GraphDataProvider.java]: Shall we move strings like OS, JDK, DBEngine, and DBEngineVersion to constants? I believe these values will be used everywhere. Probably, we can do it later as an improvement.

File: reporting/src/main/java/org/wso2/testgrid/reporting/ChartGenerator.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
+package org.wso2.testgrid.reporting;
```
Comment (kasunbg) [inline reporting/src/main/java/org/wso2/testgrid/reporting/ChartGenerator.java]: missing license.

File: reporting/src/main/java/org/wso2/testgrid/reporting/ChartGenerator.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+        ObservableList<PieChart.Data> pieChartData =
+                FXCollections.observableArrayList(
+                        new PieChart.Data("Failed", failedCount),
+                        new PieChart.Data("Skipped", skippedCount),
+                        new PieChart.Data("Passed", passedCount));
+        final PieChart chart = new PieChart(pieChartData);
+        chart.setAnimated(false);
+        chart.setTitle("Build Failure Summary by Test Plans");
+        genChart(chart, 500, 500, summaryChartFileName);
```
Comment (kasunbg) [inline reporting/src/main/java/org/wso2/testgrid/reporting/ChartGenerator.java]: What are 500s here?

File: reporting/src/main/java/org/wso2/testgrid/reporting/ChartGenerator.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        // Set Axis Names
+        xAxis.setLabel("Build Timestamp");
+        yAxis.setLabel("Number of Infra Combinations");
+
+        // Setting series names
+        seriesSet[0].setName("Build Failed Combinations");
+        seriesSet[1].setName("Build Passed Combinations");
+        seriesSet[2].setName("Infra Failed Combinations");
+        // Setting space between the bars
+        stackedBarChart.setCategoryGap(50);
```
Comment (kasunbg) [inline reporting/src/main/java/org/wso2/testgrid/reporting/ChartGenerator.java]: From where did this magic number come from? :)

File: reporting/src/main/java/org/wso2/testgrid/reporting/TestReportEngine.java [hunk 8]
```diff
@@ -813,4 +823,121 @@ private void postProcessSummaryTable(Map<String, InfrastructureBuildStatus> test
...
+        } catch (TestGridException e) {
+            throw new ReportingException(
+                    "Error occurred while getting the test plan id by reading test grid yaml files ", e);
+        }
+    return testPlans;
+    }
+
+    /**
+     * Generate summarized charts for the summary email.
+     *  @param workspace curent workspace of the build
```
Comment (harshanL) [inline reporting/src/main/java/org/wso2/testgrid/reporting/TestReportEngine.java]: Unnecessary whitespace here.

File: reporting/src/main/java/org/wso2/testgrid/reporting/ChartGenerator.java (hunk 6 above)
Comment (yasassri) [reply reporting/src/main/java/org/wso2/testgrid/reporting/ChartGenerator.java]: This is the size of the generated graph.

File: reporting/src/main/java/org/wso2/testgrid/reporting/ChartGenerator.java (hunk 7 above)
Comment (yasassri) [reply reporting/src/main/java/org/wso2/testgrid/reporting/ChartGenerator.java]: This is an aesthetic improvement, so just added a number which looked nice :)

**Comment by kasunbg** [comment]
Then this needs to go to reporting module, not to common. This'll lead to
major confusion otherwise.

File: reporting/src/main/java/org/wso2/testgrid/reporting/TestReportEngine.java [hunk 9]
```diff
@@ -76,8 +82,10 @@
     private static final String REPORT_MUSTACHE = "report.mustache";
     private static final String PERFORMANCE_REPORT_MUSTACHE = "performance_report.mustache";
     private static final String EMAIL_REPORT_MUSTACHE = "email_report.mustache";
+    private static final String SUMMARIZED_EMAIL_REPORT_MUSTACHE = "summarized_email_report.mustache";
```
Comment (kasunbg) [inline reporting/src/main/java/org/wso2/testgrid/reporting/TestReportEngine.java]: We need to merge these two reports together. May be in a later PR.

Comment (yasassri) [reply reporting/src/main/java/org/wso2/testgrid/reporting/TestReportEngine.java]: +1

File: common/src/main/java/org/wso2/testgrid/common/InfraCombination.java (hunk 1 above)
Comment (kasunbg) [reply common/src/main/java/org/wso2/testgrid/common/InfraCombination.java]: Then this needs to go to reporting module, not to common. This'll lead to
major confusion otherwise.

File: web/src/main/java/org/wso2/testgrid/web/api/ProductService.java [hunk 10]
```diff
@@ -129,6 +129,8 @@ public Response getAllProductStatuses() {
                 status.setLastFailureTimestamp(product.getLastFailureTimestamp());
                 status.setProductStatus(testPlanUOW.getCurrentStatus(product).toString());
                 list.add(status);
+//                GraphDataProvider gp = new GraphDataProvider();
+//                gp.getTestExecutionHistory("1d68c7b1-210f-4794-a8a4-58e798155905");
```
Comment (kasunbg) [inline web/src/main/java/org/wso2/testgrid/web/api/ProductService.java]: commented out code.

File: common/src/main/java/org/wso2/testgrid/common/InfraCombination.java (hunk 1 above)
Comment (yasassri) [reply common/src/main/java/org/wso2/testgrid/common/InfraCombination.java]: We are using the same object for mapping the values in mustache, so shall we move this into report models?

File: reporting/src/main/java/org/wso2/testgrid/reporting/GraphDataProvider.java (hunk 4 above)
Comment (kasunbg) [reply reporting/src/main/java/org/wso2/testgrid/reporting/GraphDataProvider.java]: +1. We can keep these within reporting module. testgrid core is built in an abstract way. So, it does not know about these strings.

File: reporting/src/main/java/org/wso2/testgrid/reporting/model/email/TestCaseResultSection.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
+package org.wso2.testgrid.reporting.model.email;
```
[the remark this answers is not in the document]
Comment (lasanthaDLPDS) [reply reporting/src/main/java/org/wso2/testgrid/reporting/model/email/TestCaseResultSection.java]: This is fixed with 923905c

File: reporting/src/main/java/org/wso2/testgrid/reporting/ChartGenerator.java (hunk 5 above)
Comment (lasanthaDLPDS) [reply reporting/src/main/java/org/wso2/testgrid/reporting/ChartGenerator.java]: This is fixed with 923905c
